File: instrument_select.py
import time
from grid_select import Grid

class InstrumentSelect:
    def __init__(self, toggled_instrument):
        self.instruments = ["Bass", "Bassoon", "Cello", "Clarinet", "Contrabass",
                            "Flute", "French Horn", "Acoustic Guitar", "Electric Guitar", "Classical Guitar",
                            "Harmonium", "Harp", "Organ", "Piano", "Saxophone", "Trombone", "Trumpet","Tuba",
                            "Violin", "Xylophone", "AMSynth", "FMSynth", "MembraneSynth","DuoSynth"]
        self.width = 6
        self.grid = Grid(self.instruments, self.width)
        self.currInstrument = ""
        self.prevTriggered = True
        self.toggledInstrument = toggled_instrument

    def instrumentNotification(self, direction, triggered):
        if (triggered and self.prevTriggered):
            self.prevTriggered = False
            self.toggledInstrument = self.currInstrument
        
            return (self.currInstrument, True, True)

        newX, newY = self.grid.moveValue(direction)
        newInstrument = self.currInstrument
        if (self.grid.isValid(newX, newY, []) == True):
            # self.x, self.y = newX, newY
            self.grid.move(newX, newY)
            self.currInstrument = self.grid._grid[newY][newX]
            
        elif (self.grid.isValid(newX, newY, []) == 'skip'):
            
            if direction == 'up':
                newY -= 1
            elif direction == 'down':
                newY += 1
            elif direction == 'right':
                newX += 1
            elif direction == 'left':
                newX -= 1

            if (self.grid.isValid(newX, newY, [])):
                self.grid.move(newX, newY)
                self.currInstrument = self.grid._grid[newY][newX]

        if (newInstrument != self.currInstrument):
            self.prevTriggered = True
            return (self.currInstrument, False, True)
        else:
            return (self.currInstrument, False, False)

File: static/js/InstrumentSelect.js
const instruments = ["Bass", "Bassoon", "Cello", "Clarinet", "Contrabass",
    "Flute", "French Horn", "Acoustic Guitar", "Electric Guitar", "Classical Guitar",
    "Harmonium", "Harp", "Organ", "Piano", "Saxophone", "Trombone", "Trumpet","Tuba",
    "Violin", "Xylophone","AMSynth", "FMSynth", "MembraneSynth","DuoSynth"];

class InstrumentSelect {
    constructor () {
        socket.on('send instrument', (msg) => {
            this.setInstrument(msg);
        });
        this.currInstrument = 'Bass'
        this.selectedInstrument = 'Bass'
    }

    generateInstruments = () => {
        var instrumentRow = document.getElementById("instrument-list");
        for (var i = 0; i < instruments.length; i++) {
            const instrumentName = instruments[i];
            var col = document.createElement("div");
            col.className = 'col-2';

            var card = document.createElement("div");
            card.id = instrumentName;
            card.className = 'card mb-4 shadow-sm instrument-card';

            var para = document.createElement("p");
            para.id = 'instrument-text';
            para.innerText = instrumentName;

            card.appendChild(para);
            col.appendChild(card);
            instrumentRow.appendChild(col);
        }
    }

    setInstrument = (data) => {
        var instrumentName = data['instrument']
        var triggered = data['change']
        
        if (instrumentName != this.currInstrument) {
            console.log(this.currInstrument);
            var instrumentCard = document.getElementById(this.currInstrument);
            instrumentCard.classList.remove('highlighted');
            this.currInstrument = instrumentName;
            
            instrumentCard = document.getElementById(this.currInstrument);
            instrumentCard.classList.add('highlighted');
        }

        if (triggered) {
            var instrumentCard = document.getElementById(this.selectedInstrument);
            instrumentCard.classList.remove('selected');
            this.selectedInstrument = instrumentName;

            instrumentCard = document.getElementById(this.selectedInstrument);
            instrumentCard.classList.add('selected')
        }
    }
}